phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -927,109 +1123,375 @@ private void verifyAndOrRebuildIndex() throws IOException {
+            for (Cell cell : cells) {
+                switch (cell.getType()) {
+                    case DeleteFamily:
+                    case DeleteFamilyVersion:
+                        newDel.addDeleteMarker(cell);
+                }
+            }
+        }
+        return newDel;
+    }
+
+    public static Delete adjustGlobalIndexDeleteMutation(Delete del) throws IOException {
+        // For a global index table, we are only interested if the current data row is deleted or
+        // not. There is no need to apply column deletes to index rows since index rows are always
+        // full rows and all the cells in an index row have the same timestamp value. Because of
+        // this index rows versions do not share cells.
+        boolean deleteColumn = false;
+        boolean deleteFamily = false;
+        for (List<Cell> cells : del.getFamilyCellMap().values()) {
+            for (Cell cell : cells) {
+                switch (cell.getType()) {
+                    case DeleteFamily:
+                    case DeleteFamilyVersion:
+                        deleteFamily = true;
+                        break;
+                    case DeleteColumn:
+                    case Delete:
+                        deleteColumn = true;
+                }
+            }
+        }
+        if (!deleteFamily) {
+            return null;
+        }
+        if (deleteColumn) {
+            del = copyDeleteFamilyMarkers(del);
+        }
+        return del;
+    }
+
+    private static void addIndexDeleteMutation(IndexMaintainer indexMaintainer, Delete del, List<Mutation> mutationList,
+                                               long ts) throws IOException {
+        // The following unverified put is needed for inline verification to treat regular updates and
+        // rebuild updates the same way
+        Put unverifiedPut = new Put(del.getRow());
+        unverifiedPut.addColumn(indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                indexMaintainer.getEmptyKeyValueQualifier(), ts, UNVERIFIED_BYTES);
+        mutationList.add(unverifiedPut);
+        del = adjustGlobalIndexDeleteMutation(del);
+        mutationList.add(del);
+    }
+
+    public static Collection<Cell> getCollection(Mutation m) {
+        Collection<Cell> collection = new ArrayList<>();
+        for (List<Cell> cells : m.getFamilyCellMap().values()) {
+            collection.addAll(cells);
+        }
+        return collection;
+    }
+
+
+    /**
+     * Generate the index update for a data row from the mutation that are obtained by merging the previous data row
+     * state with the pending row mutation for index rebuild. This method is called only for global indexes.
+     * pendingMutations is a sorted list of data table mutations that are used to replay index table mutations.
+     * This list is sorted in ascending order by the tuple of row key, timestamp and mutation type where delete comes
+     * after put.
```

Comment:
Little confused here, MUTATION_TS_COMPARATOR puts deletes before puts, comment here says otherwise?

Reply:
I have corrected the comment on MUTATION_TS_COMPARATOR.